AxonDeepSeg/ads_utils.py

```diff
@@ -254,7 +257,18 @@ def convert_path(object_path):
 def imread(filename, bitdepth=8):
     """ Read image and convert it to desired bitdepth without truncation.
     """
-    if 'tif' in str(filename):
+
+    # Convert to Path
+    filename = Path(filename)
+
+    file_ext = filename.suffix.lower()
```

Comment:
This will return ".tif" for both ".tif" files and ".ome.tif" files.
We currently don't support OME-TIF files in ADS (nor ivadomed) with imageio/pillow.
However OME-TIF is one of the standard format in BIDS so they are likely to be found in microscopy datasets.

To avoid having an OME-TIF mistaken for a TIF in ivadomed, we use an "ordered" list of extensions that returns the first extension that match the end of the filename [here](https://github.com/ivadomed/ivadomed/blob/7b76bf81a025cde3096fd1d686d6f3c0b8ce8f02/ivadomed/loader/utils.py#L404).

Reply:
In that case we should use a function like in ivadomed because this pattern of suffix extraction is everywhere.

Reply:
> ".ome.tif"

I was not aware that we had to provide support for this file extension (I hate double file extensions! but I get it).

Are there any other ones that we need to consider that are not or have never been in our valid extensions list? "ome.tiff", "ome.png", etc?

Reply:
> I was not aware that we had to provide support for this file extension (I hate double file extensions! but I get it).

Yeah, double extensions are not fun.
The implementation in this PR works as intended with what we discussed before, I just thought about that corner case yesterday while reviewing it.

In ivadomed we don't support "ome" files either at the moment but having them in the list of extension is a way to identify them correctly before they enter the "traditional tif" pipeline. If the file is identified as an "ome", it currently raise an error [here](https://github.com/ivadomed/ivadomed/blob/7b76bf81a025cde3096fd1d686d6f3c0b8ce8f02/ivadomed/loader/segmentation_pair.py#L284).

> Are there any other ones that we need to consider that are not or have never been in our valid extensions list? ".ome.tiff", ".ome.png", etc?

We could use the same list as ivadomed [here](https://github.com/ivadomed/ivadomed/blob/7b76bf81a025cde3096fd1d686d6f3c0b8ce8f02/ivadomed/loader/utils.py#L35) except `.nii` and `.nii.gz`.

Reply:
@mariehbourget I've resolved this issue you raised in commits [2fe92b7](https://github.com/axondeepseg/axondeepseg/pull/629/commits/2fe92b723e9e45a531257643d05ac759ac904cf2)  [084a386](https://github.com/axondeepseg/axondeepseg/pull/629/commits/084a386f293fa83894a823fcfbe4af0c8ca3d025) [5cb822a](https://github.com/axondeepseg/axondeepseg/pull/629/commits/5cb822a41df3d66c3b4a938229d3b6162af08ec2)

In particular, here is a working example of the filename extension extraction I implemented in these lines

https://github.com/axondeepseg/axondeepseg/blob/5cb822a41df3d66c3b4a938229d3b6162af08ec2/AxonDeepSeg/ads_utils.py#L261-L266


```python
Python 3.9.12 (main, Mar 26 2022, 15:51:15) 
[Clang 13.1.6 (clang-1316.0.21.2)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from pathlib import Path
>>> filename = Path('foo.OME.tif')
>>> print(''.join(filename.suffixes).lower())
.ome.tif
>>> 
```

Reply:
I completely missed the notification 13 days ago 😬, my apologies for the delay.

At first glance, I really like this method compared to what I have done in ivadomed. Unfortunately, I think I found a bug here where:
if there are "dots" in the filename, they will be picked up as file extensions, ex:
```
>>> filename = Path('foo.bar.tif')
>>> print(''.join(filename.suffixes).lower())
.bar.tif
```
This unfortunately happens often with non-BIDS filenames containing all sorts of illegal characters like:
`mouse Nerve-1.4NA-63x.tif`.

Let me know your thoughts.

Reply:
>This unfortunately happens often with non-BIDS filenames containing all sorts of illegal characters like:
mouse Nerve-1.4NA-63x.tif.

>Let me know your thoughts.

My thoughts are we shouldn't support filenames with dots in them 😅 , what a mess that would cause 😬. Dots in filenames are meant to delineate the suffixes, so if a user came to me with a dataset like that, at most I'd suggest a script that renames it (replacing the non-suffix related dot with a dash).

But that's my view - do we have many datasets already like that? If it's an edge case (eg, one user), I don't think we should support it.

Otherwise, I'm open to a suggestion as to how to extract suffixes *multiple dots* (i.e. `.ome.tiff`) to check if they are in the filename, but not random non-suffix dots (i.e. your `.4NA-63x.tif` example above). We can't just parse the list of extensions and check if they are in the filenames (eg, `is 'ome.tiff' in filename`), because that's what caused the bug in the first place (e.g, `is tif in filename ` will be true for something like `tiffany.txt`)

Reply:
> But that's my view - do we have many datasets already like that? If it's an edge case (eg, one user), I don't think we should support it.

Yeah my example came from a real user. So it happened in the past but I don't know to which extent.

> Otherwise, I'm open to a suggestion as to how to extract suffixes _multiple dots_ (i.e. `.ome.tiff`) to check if they are in the filename, but not random non-suffix dots (i.e. your `.4NA-63x.tif` example above). We can't just parse the list of extensions and check if they are in the filenames (eg, `is 'ome.tiff' in filename`), because that's what caused the bug in the first place (e.g, `is tif in filename ` will be true for something like `tiffany.txt`)

Here is how we do it in ivadomed which seems to work: https://github.com/ivadomed/ivadomed/blob/b9a208d79254abce847ae9d522fea376057ce2b3/ivadomed/loader/utils.py#L408

But we would have to add an error, as @hermancollin suggested if the extension ends up being some sort of "ome" because we don't support it. (same for ivadomed: https://github.com/ivadomed/ivadomed/blob/b9a208d79254abce847ae9d522fea376057ce2b3/ivadomed/loader/segmentation_pair.py#L283)

Reply:
@mariehbourget thanks a lot for these resources! I've integrated them into our utils file and modified them slightly, and wrote tests for them as well. Please let me know if with these changes you think this PR is ready!